internal/cron/verify_domain_ownership.go

```diff
@@ -0,0 +1,88 @@
+package cron
+
+import (
+	"context"
+	"errors"
+	"net"
+
+	"github.com/oursky/pageship/internal/db"
+	"github.com/oursky/pageship/internal/models"
+	"github.com/oursky/pageship/internal/time"
+	"go.uber.org/zap"
+)
+
+type VerifyDomainOwnership struct {
+	Schedule                     string
+	DB                           db.DB
+	MaxConsumeActiveDomainCount  uint
+	MaxConsumePendingDomainCount uint
+}
+
+func (v *VerifyDomainOwnership) Name() string { return "verify-domain-ownership" }
+
+func (v *VerifyDomainOwnership) CronSchedule() string { return v.Schedule }
+
+func (v *VerifyDomainOwnership) Run(ctx context.Context, logger *zap.Logger) error {
+	now := time.SystemClock.Now().UTC()
+
+	return db.WithTx(ctx, v.DB, func(c db.Tx) error {
+		verified := true
+		unverified := false
+		domainVerifications := []*models.DomainVerification{}
+		verifiedRecords, _ := c.ListDomainVerifications(ctx, nil, &v.MaxConsumeActiveDomainCount, &verified)
+		unverifedRecords, _ := c.ListDomainVerifications(ctx, nil, &v.MaxConsumePendingDomainCount, &unverified)
+		for _, verifiedRecord := range verifiedRecords {
+			domainVerifications = append(domainVerifications, verifiedRecord)
+		}
+		for _, unverifiedRecord := range unverifedRecords {
+			domainVerifications = append(domainVerifications, unverifiedRecord)
+		}
+
+		var consumedCount = 0
+		for _, domainVerification := range domainVerifications {
+			key, value := domainVerification.GetTxtRecord()
+			values, err := net.LookupTXT(key)
+			var dnsErr *net.DNSError
+			if err != nil && errors.As(err, &dnsErr) && !dnsErr.IsNotFound {
+				continue
+			}
+			consumedCount += 1
+			matched := false
+			for _, v := range values {
+				if v == value {
+					matched = true
+					break
+				}
+			}
+			domain, err := c.GetDomainByName(ctx, domainVerification.Domain)
+			if matched {
+				domainVerification.VerifiedAt = &now
+				if errors.Is(err, models.ErrDomainNotFound) {
+					domainName := domainVerification.Domain
+					app, err := c.GetApp(ctx, domainVerification.AppID)
+					if err != nil {
+						continue
+					}
+					config, ok := app.Config.ResolveDomain(domainName)
+					if ok {
+						c.CreateDomain(ctx, models.NewDomain(
+							now, domainName, domainVerification.AppID, config.Site,
+						))
+					}
+				}
+			} else {
+				domainVerification.VerifiedAt = nil
+				if domain != nil {
+					c.DeleteDomain(ctx, domain.ID, now)
+				}
+			}
+			domainVerification.UpdatedAt = now
```

Comment:
I'm feeling reserved on using `UpdatedAt` as trigger for verification; updates could be caused by operations other than domain verification requests.

Also seems that if verification failed, the flow is stuck unless user delete the domain and re-add it.

I would suggest to add two columns for this: `last_checked_at` and `will_check_at`.
- The cron job lookup N expired verification requests (i.e. `will_check_at` is before now)
- Update `last_checked to be now after checking.
- If check passed: update `will_check_at` to `now + X` (X could be arbitrary value, like 1h).
- If check failed: update `will_check_at` to be null to stop cron job from checking it.
- If user requested check, set `will_check_at` to now.
- Display `last_checked_at` to user to show that the server is indeed running the check process.

Reply:
OH, i thought it validate every minutes or we can put X in app config then only `last_checked_at` is required as `will_check_at` is derived value.

Reply:
Not sure how would it handle the 4th/5th point. If just `last_checked_at` is used, how to distinguish between a failed check and pending check, and allow user to retry the check?

Reply:
You are right. The two fields are essentials for saving worker from busy.